File: java/app/src/main/java/org/example/tds/td3/promowarn/version3/Student.java
package org.example.tds.td3.promowarn.version3;

import org.example.tds.td3.promowarn.common.mail.EMailAddress;

import java.util.Optional;

public interface Student {
    Integer id();

    EMailAddress email();

    Optional<Double> grade();

    void grade(final double grade);
}

File: scala/src/main/scala/tds/td3/api/Student.scala
package tds.td3.api

import tds.td3.common.mail.EMailAddress

// NON FUNCTIONAL: MUTATION
trait Student extends Identified:

  val email: EMailAddress

  def grade: Option[Double]

  // NON FUNCTIONAL: MUTATION
  def setGrade(grade: Double): Unit
